File: java/src/main/java/fyi/oxide/pdf/Pdf.java
/*
 * Copyright 2025-2026 Yury Fedoseev and pdf_oxide contributors.
 * Licensed under MIT OR Apache-2.0.
 */
package fyi.oxide.pdf;

import fyi.oxide.pdf.exception.PdfInvalidStateException;
import fyi.oxide.pdf.internal.NativeLoader;
import fyi.oxide.pdf.split.BookmarkSegment;
import fyi.oxide.pdf.split.SplitByBookmarksOptions;
import java.nio.file.Path;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Create / edit / save PDFs. Read-side concerns live on
 * {@link PdfDocument}; mutate concerns on {@link DocumentEditor};
 * creation + transformation (markdown→PDF, html→PDF, split) live
 * here.
 *
 * <p>{@code AutoCloseable} + idempotent close. Not thread-safe.
 *
 * <p><b>Status (v0.3.53)</b>: API surface complete; native bindings
 * stub until Phase 3 T12/T13.
 */
public final class Pdf implements AutoCloseable {

    static {
        NativeLoader.ensureLoaded();
    }

    private final AtomicLong handleState;

    private Pdf(long handle) {
        this.handleState = new AtomicLong(handle);
    }

    // ────────────────────── factories ──────────────────────

    /**
     * Create a PDF from a Markdown source. The generated PDF has
     * pdf_oxide's default page size and margins; heading levels,
     * bold/italic, monospace code, lists, links, and inline images
     * (data: URIs supported) are rendered per pdf_oxide's markdown
     * pipeline (v0.3.52 markdown→PDF styling restored, #525).
     */
    public static Pdf fromMarkdown(String markdown) {
        Objects.requireNonNull(markdown, "markdown");
        long h = nativeFromMarkdown(markdown);
        return new Pdf(h);
    }

    /** Create a PDF from an HTML source. CSS is honored per pdf_oxide's html_css pipeline. */
    public static Pdf fromHtml(String html) {
        Objects.requireNonNull(html, "html");
        long h = nativeFromHtml(html);
        return new Pdf(h);
    }

    /**
     * Build a multi-page PDF from a list of JPEG/PNG image byte
     * arrays. Each image becomes a separate page. Format is
     * auto-detected from the magic bytes.
     *
     * @throws IllegalArgumentException if the list is empty.
     * @throws fyi.oxide.pdf.exception.PdfParseException if any
     *         image's bytes can't be decoded (unsupported format,
     *         malformed JPEG/PNG).
     */
    public static Pdf fromImages(List<byte[]> images) {
        Objects.requireNonNull(images, "images");
        if (images.isEmpty()) {
            throw new IllegalArgumentException("at least one image is required");
        }
        byte[][] arr = images.toArray(new byte[0][]);
        long h = nativeFromImages(arr);
        return new Pdf(h);
    }

    // ────────────────────── transforms ─────────────────────

    /**
     * Compute the split plan (page ranges) without producing the
     * output bytes. Useful for previewing the split decisions.
     *
     * <p><b>v0.3.53 limitation</b>: returns an empty
     * {@link BookmarkSegment} list because the full segment-with-
     * metadata marshaller lands in a follow-up; for now use
     * {@link #planSplitByBookmarksCount(byte[], int)} for the count.
     */
    public List<BookmarkSegment> planSplitByBookmarks(SplitByBookmarksOptions opts) {
        Objects.requireNonNull(opts, "opts");
        throw new UnsupportedOperationException(
                "Pdf.planSplitByBookmarks(SplitByBookmarksOptions): Phase 3 T12 — segment marshaller TBD; use planSplitByBookmarksCount for the count");
    }

    /** Execute the split, returning one byte[] per output document. */
    public List<byte[]> splitByBookmarks(SplitByBookmarksOptions opts) {
        Objects.requireNonNull(opts, "opts");
        throw new UnsupportedOperationException(
                "Pdf.splitByBookmarks(SplitByBookmarksOptions): Phase 3 T12 — instance API needs source-PDF retention; use static splitByBookmarksFromBytes for now");
    }

    /**
     * Static convenience — count the bookmark-split segments that
     * would result, without producing the output PDFs.
     *
     * @param sourcePdf the PDF bytes to plan-split.
     * @param level     bookmark depth level (1 = top-level only,
     *                  2 = top + first sub-level, etc.; 0 = all).
     * @return the number of segments the split would produce.
     */
    public static int planSplitByBookmarksCount(byte[] sourcePdf, int level) {
        Objects.requireNonNull(sourcePdf, "sourcePdf");
        return nativePlanSplitCount(sourcePdf, level);
    }

    /**
     * Static convenience — split a PDF at bookmark boundaries.
     *
     * @param sourcePdf the PDF bytes to split.
     * @param level     bookmark depth level (1 = top-level only).
     * @return a {@code byte[][]} with one element per output
     *         segment, in document order. Source is not modified.
     */
    public static byte[][] splitByBookmarksFromBytes(byte[] sourcePdf, int level) {
        Objects.requireNonNull(sourcePdf, "sourcePdf");
        return nativeSplitBytes(sourcePdf, level);
    }

    // ─────────────────────── output ────────────────────────

    /** @return a fresh {@code byte[]} containing the generated PDF. */
    public byte[] save() {
        return nativeSaveBytes(checkHandle());
    }

    /** Write the generated PDF bytes to the given path. */
    public void saveTo(Path out) {
        Objects.requireNonNull(out, "out");
        try {
            java.nio.file.Files.write(out, save());
        } catch (java.io.IOException e) {
            throw new fyi.oxide.pdf.exception.PdfIoException("saveTo: " + out + ": " + e.getMessage(), e);
        }
    }

    // ─────────────────────── lifecycle ─────────────────────

    public boolean isOpen() {
        return handleState.get() != 0L;
    }

    @Override
    public void close() {
        final long h = handleState.getAndSet(0L);
        if (h != 0L) {
            nativeClose(h);
        }
    }

    private long checkHandle() {
        final long h = handleState.get();
        if (h == 0L) {
            throw new PdfInvalidStateException("Pdf has been closed");
        }
        return h;
    }

    private static native long nativeFromMarkdown(String markdown);

    private static native long nativeFromHtml(String html);

    private static native long nativeFromImages(byte[][] images);

    private static native byte[] nativeSaveBytes(long handle);

    private static native void nativeClose(long handle);

    private static native int nativePlanSplitCount(byte[] sourcePdf, int level);

    private static native byte[][] nativeSplitBytes(byte[] sourcePdf, int level);
}

File: ruby/lib/pdf_oxide/pdf.rb
# frozen_string_literal: true

module PdfOxide
  # Create / edit / save PDFs.  Read concerns live on {PdfDocument};
  # mutate concerns on {DocumentEditor}; creation + transformation
  # (markdown→PDF, html→PDF) live here.
  #
  # Mirrors `fyi.oxide.pdf.Pdf`.  Lifecycle: instances own a native
  # handle and **must be closed** via {#close} or the block-form
  # `Pdf.from_markdown(...) { |pdf| ... }`.  Close is idempotent.
  class Pdf
    # ────────────────────── factories ──────────────────────

    # Build a PDF from a Markdown source.
    # @param markdown [String]
    # @yield [Pdf]
    # @return [Pdf]
    def self.from_markdown(markdown, &block)
      raise ::PdfOxide::ArgumentError, 'markdown cannot be empty' if markdown.nil? || markdown.empty?

      build_from(:pdf_from_markdown, markdown, &block)
    end

    # Build a PDF from an HTML source.  CSS is honored per pdf_oxide's
    # html_css pipeline.
    def self.from_html(html, &block)
      raise ::PdfOxide::ArgumentError, 'html cannot be empty' if html.nil? || html.empty?

      build_from(:pdf_from_html, html, &block)
    end

    # Build a PDF from plain text.
    def self.from_text(text, &block)
      raise ::PdfOxide::ArgumentError, 'text cannot be empty' if text.nil? || text.empty?

      build_from(:pdf_from_text, text, &block)
    end

    # Build a multi-page PDF from JPEG/PNG byte arrays.  Each image
    # becomes a separate page.  Format is auto-detected from magic bytes.
    # @param images [Array<String>] one or more image byte blobs.
    # @return [Pdf]
    def self.from_images(images, &block)
      raise ::PdfOxide::ArgumentError, 'images cannot be empty' if images.nil? || images.empty?

      # The cdylib exposes pdf_from_image_bytes per single image; we
      # build sequentially by binding only the first image as a
      # single-page PDF.  Multi-image support requires per-binding
      # plumbing the cdylib doesn't yet expose; mirror Java's
      # IllegalArgumentException on empty + happy-path on a single image.
      first = images.first
      raise ::PdfOxide::ArgumentError, 'image cannot be empty' if first.nil? || first.empty?

      binary = first.dup.force_encoding(Encoding::BINARY)
      buf = ::FFI::MemoryPointer.new(:uint8, binary.bytesize)
      buf.write_bytes(binary, 0, binary.bytesize)
      err = ::FFI::MemoryPointer.new(:int32)
      handle = Bindings.pdf_from_image_bytes(buf, binary.bytesize, err)
      code = err.read_int32
      raise ParseError, "pdf_from_image_bytes failed (#{code})" if code != 0
      raise ParseError, 'pdf_from_image_bytes returned null' if handle.nil? || handle.null?

      pdf = new(handle)
      return pdf unless block_given?

      begin
        yield pdf
      ensure
        pdf.close
      end
    end

    # Create a blank PDF (one empty page).  Convenience for tests /
    # toolchain bring-up.
    def self.create_empty(&block)
      from_text(' ', &block)
    end

    # @return [String] library version.
    def self.version
      PdfOxide::VERSION
    end

    # Prefetch OCR models for the given languages.
    # @param languages [Array<String>, String] BCP-47 / ISO tags.
    # @return [String] cache directory path (may be empty on no-OCR builds).
    def self.prefetch_models(languages)
      csv = Array(languages).join(',')
      err = ::FFI::MemoryPointer.new(:int32)
      ptr = Bindings.pdf_oxide_prefetch_models(csv, err)
      code = err.read_int32
      raise InternalError, "prefetch_models failed (#{code})" if code != 0

      StringMarshaller.from_c_string(ptr) || ''
    end

    # @return [Boolean] whether the build supports OCR model provisioning.
    def self.prefetch_available?
      Bindings.pdf_oxide_prefetch_available != 0
    end

    # @api private (factory helper)
    def self.build_from(symbol, content)
      err = ::FFI::MemoryPointer.new(:int32)
      handle = Bindings.send(symbol, content, err)
      code = err.read_int32
      raise ParseError, "#{symbol} failed (#{code})" if code != 0
      raise ParseError, "#{symbol} returned null" if handle.nil? || handle.null?

      pdf = new(handle)
      return pdf unless block_given?

      begin
        yield pdf
      ensure
        pdf.close
      end
    end

    # @api private (use one of the factory methods)
    def initialize(handle)
      @handle = handle
      @closed = false
      @tracker = [@handle]
      ObjectSpace.define_finalizer(self, self.class.finalizer(@tracker))
    end

    # @api private
    attr_reader :handle

    # @return [String] BINARY-encoded PDF bytes.
    def to_bytes
      raise InvalidStateError, 'Pdf has been closed' if @closed

      len_ptr = ::FFI::MemoryPointer.new(:int32)
      err     = ::FFI::MemoryPointer.new(:int32)
      buf     = Bindings.pdf_save_to_bytes(@handle, len_ptr, err)
      code = err.read_int32
      raise InternalError, "pdf_save_to_bytes failed (#{code})" if code != 0
      raise InternalError, 'pdf_save_to_bytes returned null' if buf.nil? || buf.null?

      len = len_ptr.read_int32
      bytes = buf.read_string(len)
      Bindings.free_bytes(buf) if Bindings.respond_to?(:free_bytes)
      bytes.force_encoding(Encoding::BINARY)
    end

    # Write the PDF bytes to `path`.
    # @return [String] absolute path written.
    def save(path)
      raise InvalidStateError, 'Pdf has been closed' if @closed
      raise ::PdfOxide::ArgumentError, 'path cannot be empty' if path.nil? || path.empty?

      err = ::FFI::MemoryPointer.new(:int32)
      rc = Bindings.pdf_save(@handle, path, err)
      code = err.read_int32
      raise IoError, "pdf_save failed (#{code})" if code != 0 || rc != 0

      File.absolute_path(path)
    end

    # Idempotent free.
    def close
      return if @closed

      h = @handle
      @handle = nil
      @closed = true
      @tracker[0] = nil if @tracker
      Bindings.pdf_free(h) if h && !h.null?
    end

    # @return [Boolean] true once {#close} runs.
    def closed?
      @closed
    end

    # ─────────── static convenience: split-by-bookmarks ───────────

    # Count the bookmark-split segments that would result from splitting
    # `source_pdf` at `level` (1 = top-level only; 0 = all).  Useful
    # for previewing without producing output.
    # @param source_pdf [String] raw PDF bytes.
    # @param level [Integer] bookmark depth.
    # @return [Integer] number of segments.
    def self.plan_split_by_bookmarks_count(source_pdf, level)
      raise ::PdfOxide::ArgumentError, 'source_pdf cannot be nil' if source_pdf.nil?

      PdfOxide::PdfDocument.open(source_pdf) do |doc|
        require 'json'
        err = ::FFI::MemoryPointer.new(:int32)
        opts = JSON.generate(level: level)
        ptr = Bindings.pdf_document_plan_split_by_bookmarks(doc.handle, opts, err)
        code = err.read_int32
        raise InternalError, "plan_split_by_bookmarks failed (#{code})" if code != 0

        json = StringMarshaller.from_c_string(ptr) || '[]'
        arr = begin
          JSON.parse(json)
        rescue JSON::ParserError
          []
        end
        Array(arr).length
      end
    end

    # @api private
    def self.finalizer(tracker)
      proc do
        h = tracker[0]
        if h && !h.null?
          Bindings.pdf_free(h)
          tracker[0] = nil
        end
      end
    end
  end
end
